### DISCUSSION TITLE
xmake v2.9.5

### DISCUSSION THREAD

**Discussion by waruqi** [xmake-io/xmake#5617] [Announcements]
## New features

* [#5462](https://github.com/xmake-io/xmake/pull/5462): Add `xmake l cli.bisect`
* [#5488](https://github.com/xmake-io/xmake/pull/5488): Support for using cosmocc to build xmake binary
* [#5491](https://github.com/xmake-io/xmake/pull/5491): Provide single xmake binary with embedded lua files
* [#5580](https://github.com/xmake-io/xmake/issues/5580): Add `@builtin/xrepo` to add envs for `xrepo env shell`

## Changes

* [#5507](https://github.com/xmake-io/xmake/issues/5507): Use treeless to improve git.clone
* [#5536](https://github.com/xmake-io/xmake/pull/5536): Add jar generate in swig mode
* [#5573](https://github.com/xmake-io/xmake/issues/5573): Improve vsxmake generator performance
* [#5601](https://github.com/xmake-io/xmake/issues/5601): Improve utils.symbols.export_all rule to filter sourcefile path

## Bugs fixed

* [#4750](https://github.com/xmake-io/xmake/issues/4750): Fix compile_commands generator for `xmake tests`
* [#5465](https://github.com/xmake-io/xmake/pull/5465): Fix lock package requires
* [#4760](https://github.com/xmake-io/xmake/issues/4760): Fix distcc client issue

### 新特性

* [#5462](https://github.com/xmake-io/xmake/pull/5462): 添加 `xmake l cli.bisect`
* [#5488](https://github.com/xmake-io/xmake/pull/5488): 支持使用 cosmocc 去构建 xmake 自身二进制
* [#5491](https://github.com/xmake-io/xmake/pull/5491): 支持提供内嵌 lua 文件的单个 xmake 二进制文件
* [#5580](https://github.com/xmake-io/xmake/issues/5580): 添加 `@builtin/xrepo` 辅助模块，为 `xrepo env shell` 实现快速设置环境变量

### 改进

* [#5507](https://github.com/xmake-io/xmake/issues/5507): 改进 git clone 下载速度
* [#5536](https://github.com/xmake-io/xmake/pull/5536): 在 swig 模式中添加 jar 生成支持
* [#5573](https://github.com/xmake-io/xmake/issues/5573): 改进 vsxmake generator 性能
* [#5601](https://github.com/xmake-io/xmake/issues/5601): 改进 utils.symbols.export_all 规则去过滤源文件路径

### Bugs 修复

* [#4750](https://github.com/xmake-io/xmake/issues/4750): 修复 compile_commands 生成器，支持 `xmake tests`
* [#5465](https://github.com/xmake-io/xmake/pull/5465): 修复 package requires lock
* [#4760](https://github.com/xmake-io/xmake/issues/4760): 修复 distcc 分布式编译问题

## What's Changed
* fixbug: for cases when WindowsSDKVersion variable is not available by @jinke18 in https://github.com/xmake-io/xmake/pull/5411
* fixbug:nsis pack only support ascii path without blank space by @jinke18 in https://github.com/xmake-io/xmake/pull/5414
* improve pch filename for msvc #5413 by @waruqi in https://github.com/xmake-io/xmake/pull/5435
* fix protobuf objectfiles #5426 by @waruqi in https://github.com/xmake-io/xmake/pull/5437
* Don't treat x86_64 as cross on arm64 macOS by @SirLynix in https://github.com/xmake-io/xmake/pull/5442
* add /sdl, /Zc:inline and ExceptionHandling flag support for `xmake project -k vs` by @jinke18 in https://github.com/xmake-io/xmake/pull/5416
* Revert e26ea1c31f9fd0733590c36db4b6f9003fae8547 by @SirLynix in https://github.com/xmake-io/xmake/pull/5447
* find libfiles for pacman by @waruqi in https://github.com/xmake-io/xmake/pull/5446
* add clang toolchain version 19 & 20 by @c8ef in https://github.com/xmake-io/xmake/pull/5451
* fix cmake compilers #5448 by @waruqi in https://github.com/xmake-io/xmake/pull/5449
* fix(plugin/project): Generate compile_commands for tests by @24bit-xjkp in https://github.com/xmake-io/xmake/pull/5460
* Add `xmake l cli.bisect` by @waruqi in https://github.com/xmake-io/xmake/pull/5462
* fix lock repo by @waruqi in https://github.com/xmake-io/xmake/pull/5465
* improve to get pkgconf for meson by @waruqi in https://github.com/xmake-io/xmake/pull/5483
* fix packagedeps by @waruqi in https://github.com/xmake-io/xmake/pull/5484
* Support to build xmake using cosmocc by @waruqi in https://github.com/xmake-io/xmake/pull/5488
* improve flags support for cmake by @waruqi in https://github.com/xmake-io/xmake/pull/5487
* Add cosmocc ci by @waruqi in https://github.com/xmake-io/xmake/pull/5489
* fix rpath for #pull/5466 by @waruqi in https://github.com/xmake-io/xmake/pull/5498
* Add cli.archive and .xmz archive format by @waruqi in https://github.com/xmake-io/xmake/pull/5491
* Improve to check and load toolchain by @waruqi in https://github.com/xmake-io/xmake/pull/5466
* use treeless instead of depth1 by @waruqi in https://github.com/xmake-io/xmake/pull/5509
* improve treeless for git.clone by @waruqi in https://github.com/xmake-io/xmake/pull/5510
* Improve invalid path for package by @waruqi in https://github.com/xmake-io/xmake/pull/5511
* fix export_all when use export_filter on arch x86 by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5512
* fix install dylib for macos app #5515 by @waruqi in https://github.com/xmake-io/xmake/pull/5519
* use ALL_BUILD project build for msvc by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5523
* improve to run tests by @waruqi in https://github.com/xmake-io/xmake/pull/5522
* try use ninja for mingw by @waruqi in https://github.com/xmake-io/xmake/pull/5521
* fix lock packages #5532 by @waruqi in https://github.com/xmake-io/xmake/pull/5539
* fix windows wasm compilation with packages by @Chi-EEE in https://github.com/xmake-io/xmake/pull/5534
* use tar to extract .tar.gz by @waruqi in https://github.com/xmake-io/xmake/pull/5540
* Support multiple targets for package by @star-hengxing in https://github.com/xmake-io/xmake/pull/5537
* improve linker for package #5542 by @waruqi in https://github.com/xmake-io/xmake/pull/5543
* Fix vulkansdk searching for mingw platform by @yh-sb in https://github.com/xmake-io/xmake/pull/5549
* Add missing Linux platform triples in autoconf tool. by @Redbeanw44602 in https://github.com/xmake-io/xmake/pull/5548
* improve cosmocc for windows by @waruqi in https://github.com/xmake-io/xmake/pull/5553
* add client parameter to rule `utils.glsl2spv` by @milkiq in https://github.com/xmake-io/xmake/pull/5563
* fix requirekey for runtimes #5538 by @waruqi in https://github.com/xmake-io/xmake/pull/5567
* Add runtimes only if the ndk version is low enough by @A2va in https://github.com/xmake-io/xmake/pull/5565
* add build settings for conan by @waruqi in https://github.com/xmake-io/xmake/pull/5569
* improve to select runtimes by @waruqi in https://github.com/xmake-io/xmake/pull/5572
* Improve to copy file by @waruqi in https://github.com/xmake-io/xmake/pull/5570
* add jar generate in swig mode by @ririyeye in https://github.com/xmake-io/xmake/pull/5536
* fix bin path for cmake #5552 by @waruqi in https://github.com/xmake-io/xmake/pull/5579
* Fix cmake find_package on wasm by @star-hengxing in https://github.com/xmake-io/xmake/pull/5578
* add envs for xrepo #5580 by @waruqi in https://github.com/xmake-io/xmake/pull/5582
* improve clang.tidy by @waruqi in https://github.com/xmake-io/xmake/pull/5586
* optimize vsxmake render #5573 by @waruqi in https://github.com/xmake-io/xmake/pull/5588
* update upload-artifact to v4 by @waruqi in https://github.com/xmake-io/xmake/pull/5591
* fix #5589 by @Domain in https://github.com/xmake-io/xmake/pull/5590
* Improve fetch package by @waruqi in https://github.com/xmake-io/xmake/pull/5594
* improve tools.cmake for wasm by @waruqi in https://github.com/xmake-io/xmake/pull/5598
* Improve export all rule by @waruqi in https://github.com/xmake-io/xmake/pull/5608
* add install.strip_packagelibs by @waruqi in https://github.com/xmake-io/xmake/pull/5611
* add xmake bundle binary by @waruqi in https://github.com/xmake-io/xmake/pull/5612
* update lz4 by @waruqi in https://github.com/xmake-io/xmake/pull/5613
* download file using powershell by @waruqi in https://github.com/xmake-io/xmake/pull/5616

## New Contributors
* @24bit-xjkp made their first contribution in https://github.com/xmake-io/xmake/pull/5460
* @Redbeanw44602 made their first contribution in https://github.com/xmake-io/xmake/pull/5548
* @milkiq made their first contribution in https://github.com/xmake-io/xmake/pull/5563

**Full Changelog**: https://github.com/xmake-io/xmake/compare/v2.9.4...v2.9.5

<hr /><em>This discussion was created from the release <a href='https://github.com/xmake-io/xmake/releases/tag/v2.9.5'>xmake v2.9.5</a>.</em>

**Comment by Domain** [comment]
> https://github.com/xmake-io/xmake/pull/5491: 支持提供内嵌 lua 文件的单个 xmake 二进制文件

这个什么意思？是说提供可以不用安装，单一的xmake可执行文件吗？是这个吗？https://github.com/xmake-io/xmake/issues/4542

**Reply by waruqi** [reply]
嗯。这个

<img width="339" alt="image" src="https://github.com/user-attachments/assets/e9c8a072-7b59-4121-84ea-4a7426dae6be">

**Comment by lewisl** [comment]
Yay!  You have the best release and upgrade process.  No one makes it easier!
